omniledger/Contracts.md

```diff
@@ -0,0 +1,191 @@
+[Cothority](https://github.com/dedis/cothority/tree/master/README.md) ::
+[Building Blocks](https://github.com/dedis/cothority/tree/master/BuildingBlocks.md) ::
+[OmniLedger](README.md) ::
+Contracts and Instances
+
+# Contracts and Instances
+
+A contract in omniledger is similar to a smart contract in Ethereum, except that
+it is pre-compiled in the code and all nodes need to have the same version of
+the contract available in order to reach consensus.
+
+A contract can spawn new instances that are tied to another contract type. All
+instances are stored in the global state of omniledger. Every instance points
+to exactly one contract. An easy interpretation is to think of a contract as
+a class and the instance as an object instantiated from that class. Or, in go,
+as the instance being the values of the `struct` and the contract being all
+methods defined on that `struct`.
+
+## Authorizations
+
+Authorizations are handled using Darcs. Each Darc has a set of rules that define
+a pair of action / expression that need to be fulfilled to execute any instruction
+on the instances governed by that Darc.
+
+A Darc is always stored with the following `InstanceID`:
+
+`[32]byte{DarcBaseID}[32]byte{0}`
+
+All instances spawned by that Darc (except for other Darcs) will share the same
+first half of the `InstanceID`. The second half is randomly chosen by the contract
+having spawned this new instance.
+
+Given the following instruction sent by a client (some fields are omitted for
+clarity):
+
+- `InstanceID`: `[32]byte{GenesisDarc}[32]byte{0x01}`
+- `Invoke`:
+  - `Command`: `Update`
+  - `Args`: `{"Roster": NewRoster}`
+- `Signatures`: `[Sig1]`
+
+omniledger will do the following:
+
+1. find the Darc instance by looking at the first 32 bytes of the `InstanceID` given in
+the instruction, here the `GenesisDarc`, and adding 32 x 0x00 bytes
+2. create a `DarcRequest` using the `InstanceID` and the `Args`
+3. verify the request corresponds to the expression of the `Invoke_Update` rule
+in the Darc instance found in 1.
+
+## Contract Arguments
+
+A contract is immutable and has the following signature in go:
+
+```go
+type OmniLedgerContract func(coll CollectionView, tx Instruction, inCoins []Coin) (sc []StateChange, outCoins []Coin, err error)
+```
+
+Input:
+- `coll` is a read-only reference to the collection representing the global state
+of all instances
+- `tx` is the instruction sent by the client, which also holds the `InstanceID`
+pointing to the data the contract should work on
+- `inCoins` is mostly ignored for the moment, but can be used to pass around
+coins between different instructions
+
+Output:
+- `sc` is the slice of stateChanges the contract wants to apply to the global
+state. They will only be applied if all instructions in the ClientTransaction
+are valid, else they will be discarded
+- `outCoins` can be used to pass around coins
+- `err` if not nil, the contract indicates it failed, and all instructions in that
```

Comment:
Wait what? Wow, I didn't understand that before. (Just checked in the code, yup!)

What's an example use case of len(tx.Instructions) > 1, where the first instruction succeeds and then a later instruction fails?

Also: It seems like if createStateChanges executes "continue clientTransactions" on the second and latter instrctions, that states will still have any states from the previous executed instructions in it. Which seems not right? If you agree, please file an issue.

Reply:
Example use case is a transaction with the two instructions:
1. Get x coins from account
2. Spend the coins creating a document

So if 2. fails, we don't want to spend the coins, but just discard the whole transaction.

But: it is important that 2. works with the changed state of 1. (which empties the account), else you could take money out of your account multiple times.